File: self_practice/py/pso_chart.m
% PSO in MATLAB with Visualization
clear; clc;

% Parameters
min_x = -4;
max_x = 2;
n_particles = 50;
max_iteration = 50;
w = 0.7;
c1 = 1.4;
c2 = 1.2;

% Function definition: f(x) = (x+1)^2 * sin(x)
f = @(x) (x+1).^2 .* sin(x);

% Initialize particle positions and velocities randomly in [min_x, max_x]
positions = min_x + (max_x - min_x) * rand(n_particles, 1);
velocities = min_x + (max_x - min_x) * rand(n_particles, 1);

% Initialize personal best positions and values
particle_best_positions = positions;
particle_best_values = inf(n_particles, 1);

% Initialize global best
global_best_value = inf;
global_best_position = positions(1);

figure;
for iter = 1:max_iteration
    % Evaluate fitness for each particle
    fitness = f(positions);
    
    % Update each particle's personal best
    for i = 1:n_particles
        if fitness(i) < particle_best_values(i)
            particle_best_values(i) = fitness(i);
            particle_best_positions(i) = positions(i);
        end
    end
    
    % Update global best
    [min_fitness, idx] = min(fitness);
    if min_fitness < global_best_value
        global_best_value = min_fitness;
        global_best_position = positions(idx);
    end
    
    % Plot function and current particle positions
    x = linspace(min_x, max_x, 400);
    y = f(x);
    plot(x, y, 'g-', 'LineWidth', 2); hold on;
    scatter(positions, f(positions), 50, 'r', 'filled'); hold off;
    title(['Iteration ' num2str(iter)]); xlabel('x'); ylabel('f(x)'); grid on;
    pause(0.1);
    
    % Update velocities and positions
    r1 = rand(n_particles, 1);
    r2 = rand(n_particles, 1);
    velocities = w .* velocities + c1 .* r1 .* (particle_best_positions - positions) + c2 .* r2 .* (global_best_position - positions);
    positions = positions + velocities;
    
    % Apply boundary conditions
    positions(positions > max_x) = max_x;
    positions(positions < min_x) = min_x;
end

disp(['Solution: ' num2str(global_best_position) ' with value: ' num2str(global_best_value)]);

File: self_practice/py/pso_chart.py
import numpy as np
import matplotlib.pyplot as plt
from random import random

def f(x):
    return (x + 1) * (x + 1) * np.sin(x)

class Particle:
    def __init__(self, min_x, max_x, dimensions=1):
        self.min_x = min_x
        self.max_x = max_x
        self.position = self.initialize(dimensions)
        self.velocity = self.initialize(dimensions)
        self.best_position = self.position
        self.best_value = 1e10

    def move(self):
        new_position = self.position + self.velocity
        new_position = np.where(new_position > self.max_x, self.max_x, new_position)
        new_position = np.where(new_position < self.min_x, self.min_x, new_position)
        self.position = new_position

    def initialize(self, x):
        return np.array([self.min_x + (self.max_x - self.min_x) * random() for _ in range(x)])

class ParticleSwarmOptimization:
    def __init__(self, min_x, max_x, n_particles=100, max_iteration=30, w=0.7, c1=1.4, c2=1.2):
        self.n_particles = n_particles
        self.max_iteration = max_iteration
        self.particles = [Particle(min_x, max_x) for _ in range(n_particles)]
        self.best_value = 1e10
        self.best_position = self.particles[0].position
        self.w = w
        self.c1 = c1
        self.c2 = c2

    def plot_function(self):
        x = np.linspace(self.particles[0].min_x, self.particles[0].max_x, 400)
        y = f(x)
        plt.plot(x, y, label="f(x)", color="green")
        plt.xlabel("x")
        plt.ylabel("f(x)")
        plt.title("Function Plot with Particle Positions")
        plt.legend()
        plt.grid()

    def plot_particles(self, iteration):
        positions = np.array([particle.position for particle in self.particles])
        plt.scatter(positions, f(positions), color="red", marker="o", label="Particles" if iteration == 0 else "")
        plt.pause(0.1)

    def run(self):
        plt.figure(figsize=(8, 6))
        self.plot_function()

        counter = 0
        while counter < self.max_iteration:
            plt.clf()
            self.plot_function()
            self.plot_particles(counter)
            self.move_particles()
            self.set_best()
            self.set_particle_best()
            counter += 1

        plt.show()
        print('Solution: %s with value: %s' % (self.best_position, self.best_value))

    def set_particle_best(self):
        for particle in self.particles:
            particle_fitness = f(particle.position)
            if particle.best_value > particle_fitness:
                particle.best_value = particle_fitness
                particle.best_position = particle.position

    def set_best(self):
        for particle in self.particles:
            particle_fitness = f(particle.position)
            if self.best_value > particle_fitness:
                self.best_value = particle_fitness
                self.best_position = particle.position

    def move_particles(self):
        for particle in self.particles:
            new_velocity = self.w * particle.velocity + self.c1 * random() * (particle.best_position - particle.position) + self.c2 * random() * (self.best_position - particle.position)
            particle.velocity = new_velocity
            particle.move()

if __name__ == '__main__':
    algorithm = ParticleSwarmOptimization(min_x=-4, max_x=2, n_particles=50, max_iteration=50)
    algorithm.run()
